JDK-8368178 Add specialization of SequencedCollection methods to emptyList, singletonList and nCopies

amaembo opened the pull request:
Please review this small change. If you have more ideas which classes may miss specializations of SequencedCollection methods, I can add them to this PR as well.

pavelrappo commented on test/jdk/java/util/Collections/EmptyListTest.java:
```diff
@@ -0,0 +1,58 @@
+ * questions.
+ */
+
+/**
```
A jtreg comment styled as a javadoc comment is one of my pet peeves. IDE and javac (since JDK 23) treats it as a javadoc comment. For example, such a comment has unknown javadoc tags, or may be considered dangling (which is what happens in your case).

amaembo replied:
Thank you for the review. I copied the comment from another test file and was not sure whether an extra asterisk is important for jtreg. You're right, it looks like it can be omitted.

pavelrappo reviewed (commented):
It's a very sensible change; thanks! The only small, orthogonal quibble inline.

amaembo commented:
@pavelrappo re `ImmutableCollections`, I was thinking about them. One thing that bugs me is that `List.copyOf(immutableList)` doesn't copy, but `List.copyOf(immutableList.reversed())` produces a fresh copy. Probably this is not a very common scenario to optimize, though.

Re `List12.reversed()` with size == 2: probably it's better to create a fresh list like `new List12<>(e1, e0)`, like

```java
        @SuppressWarnings("unchecked")
        @Override
        public List<E> reversed() {
            return e1 == EMPTY ? this : new List12<>((E) e1, e0);
        }
```

This way, we have fewer indirections and always return the instance of `List12`, which may probably help with devirtualization sometimes. The `List12` instance size is small. With compressed OOPs it's likely the same size as `ReverseOrderListView` (two object fields, compared to object+boolean fields in `ReverseOrderListView`), and the original list could become eligible for GC. One downside of this approach is that `listOf2.reversed().reversed()` will return a fresh instance, rather than the original one. But I guess it's a rare scenario and the cost is still small. What do you think?

amaembo commented:
One more concern is that `ReverseOrderListView` is not serializable, but `List12` and `ListN` are serializable. With this change, some reversed lists will be serializable while others are not. I don't see anything in the specification regarding the serializability of the reversed lists, but such a discrepancy might be confusing for some users.

pavelrappo commented:
> @pavelrappo re `ImmutableCollections`, I was thinking about them. One thing that bugs me is that `List.copyOf(immutableList)` doesn't copy, but `List.copyOf(immutableList.reversed())` produces a fresh copy. Probably this is not a very common scenario to optimize, though.
> 
> Re `List12.reversed()` with size == 2: probably it's better to create a fresh list like `new List12<>(e1, e0)`, like
> 
> ```java
>         @SuppressWarnings("unchecked")
>         @Override
>         public List<E> reversed() {
>             return e1 == EMPTY ? this : new List12<>((E) e1, e0);
>         }
> ```
> 
> This way, we have fewer indirections and always return the instance of `List12`, which may probably help with devirtualization sometimes. The `List12` instance size is small. With compressed OOPs it's likely the same size as `ReverseOrderListView` (two object fields, compared to object+boolean fields in `ReverseOrderListView`), and the original list could become eligible for GC. One downside of this approach is that `listOf2.reversed().reversed()` will return a fresh instance, rather than the original one. But I guess it's a rare scenario and the cost is still small. What do you think?


> One more concern is that `ReverseOrderListView` is not serializable, but `List12` and `ListN` are serializable. With this change, some reversed lists will be serializable while others are not. I don't see anything in the specification regarding the serializability of the reversed lists, but such a discrepancy might be confusing for some users.

---

Your comments are good and thought-through. When dealing with classes like that, it's time to bring a microscope and be concerned with as many aspects as possible.

As for the first comment, List.of and List.copyOf do not specify the class whose instance they return, or even if and how that class overrides `reversed()`. As such, I think this `@implSpec` leaves us with no choice but to abide by the roundtrip property (i.e. `list.reversed().reversed() == list`):

> **Implementation Requirements:**
>
> The implementation in this interface returns a reverse-ordered List view. The `reversed()` method of the view returns a reference to this List.

As for serializability, it also bugs be. Unmodifiable lists are designed to be consistent and strict. Even though the spec clearly says that it's a bad idea to serialize an unkown list, if a list is serializable some of the times but not others, it might hide bugs:

> In cases where the serializability of a collection is not specified, there is no guarantee about the serializability of such collections. In particular, many [view collections](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/util/Collection.html#view) are not serializable, even if the original collection is serializable.

> default List<E> reversed()
> Returns a reverse-ordered view of this collection.

I defer this serializability issue to @stuart-marks.

pavelrappo commented:
I'll slightly refine my suggestion for List12:

```java
    @Override
    public List<E> reversed() {
        return e1 == EMPTY ? this : ReverseOrderListView.of(this, false);
    }
```

pavelrappo commented:
> Although, I now wonder why the default implementation is like that.

Perhaps even that is not a mystery. If a view has a reference to the original list anyway, why not use it?

minborg commented:
In theory, `EmptyList` could implement `SequencedCollection`. Same for map.

amaembo commented:
@minborg `EmptyList` already implements it, as any other `List`:
```java
public interface List<E> extends SequencedCollection<E> { ... }
```

For `EmptyMap` (and even `SingletonMap` and `SingletonSet`), it's possible, but it will require changing the public interface (`Collections::emptyMap` will have to return `SequencedMap`), which may produce binary compatibility issues. Probably we can invent a binary-compatible signature like this:

```java
public static final <K,V,M extends Map<K, V> & SequencedMap<K, V>> M emptyMap() { ... }
```
But it looks ugly.

amaembo commented:
@minborg as JLS 13.4.15 says,
> Changing the result type of a method, or replacing a result type with void, or replacing void with a result type, has the combined effect of deleting the old method and adding a new method with the new result type or newly void result

Changing the result type from `Map` to `SequencedMap` will modify the binary signature, so compiled classes that used this method previously will fail with `NoSuchMethodError`. That's why simply changing `Map` to `SequencedMap` is not quite possible. My trick with `M extends Map<K, V> & SequencedMap<K, V>` employs the fact that the type parameter gets erased to its first bound, which is `Map`, so such kind of return type replacement is safe (but ugly).

minborg commented:
> > @minborg as JLS 13.4.15 says,
> > > Changing the result type of a method, or replacing a result type with void, or replacing void with a result type, has the combined effect of deleting the old method and adding a new method with the new result type or newly void result
> > 
> > 
> > Changing the result type from `Map` to `SequencedMap` will modify the binary signature, so compiled classes that used this method previously will fail with `NoSuchMethodError`. That's why simply changing `Map` to `SequencedMap` is not quite possible. My trick with `M extends Map<K, V> & SequencedMap<K, V>` employs the fact that the type parameter gets erased to its first bound, which is `Map`, so such kind of return type replacement is safe (but ugly).
> 
> Thanks for that JLS reference, Tagir.
> 
> @minborg, one would think they can make a method accept wider arguments or return narrower results. After all, it matches basic subtype intuitions. Yet, one cannot do that; they will get java.lang.NoSuchMethodError from code that uses the old method. FWIW, I tripped on it myself a few times.

Yeah, I get that. Another alternative would be not to expose the type in the API, but instead let the map instance returned from the method implement `SequencedMap`. A dynamic instanceof would then potentially benefit from this. Maybe there are situations that are less good as well?

RogerRiggs commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -4943,6 +4943,11 @@ public void forEach(Consumer<? super E> action) {
 
+        @Override
+        public List<E> reversed() {
+          return this;
```
Java indentation is 4 spaces.  tnx.

stuart-marks commented:
OK, List12::getFirst and List12::getLast seem fine.

But yeah... AbstractImmutableList::reversed and List12::reversed are indeed debatable. I wrestled over this for a while. I don't think they're incorrect, but I'm also having trouble convincing myself they're cases worth optimizing. There are also weird tradeoffs.

With no override, calling reversed() on a List12 returns a ReverseOrderListView(orig) and reversed() on that returns orig again. But with the override, calling reversed() gives a new List12, and calling reversed() on that gives another new List12. Nothing here is really terrible, but it's also not clear to me this is an improvement.

With AbstractImmutableList I got kind of tangled up in analyzing where this would be used. It seems to me the optimization would take effect for the zero-size nulls-disallowed case, the zero-size nulls-allowed case, and the size-one nulls-allowed case. (The size-one nulls-disallowed case is covered by List12, unless we don't add the override there....) I don't think this is incorrect for any case, but I'm also having a hard time seeing the improvement. It might save the allocation of a small object, or it might save an indirection. And ... that's about it?

The reversed() overrides hardly seem worthwhile. Maybe don't include them.

Regarding the tests, is it necessary to add new test files for EmptyList and SingletonList? Maybe it's possible to add cases to test/jdk/java/util/Collection/MOAT.java. I'm also a bit surprised to see a new provider added to the ListFactories test. But I haven't looked too closely. If it's necessary, then of course we should include it.

amaembo commented:
> The reversed() overrides hardly seem worthwhile. Maybe don't include them.

Thank you. Agreed. I removed them.

> Maybe it's possible to add cases to test/jdk/java/util/Collection/MOAT.java

It's possible, but these old-style tests look fragile and less maintainable. Nevertheless, agreed, this is more consistent. I added a few new cases to MOAT.java and removed the new test classes.

>  I'm also a bit surprised to see a new provider added to the ListFactories test.

I thought that we may test reversed() views thoroughly in the same way as subList() views are tested. But without reversed specializations, it looks like it's not so necessary. So I removed this provider.
